rfd/0056-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,702 @@
+`tctl` will ensure that:
+
+- the objects related to a bot all follow a standardized naming convention that
+  includes the name of the bot
+- a new `teleport.dev/owned-by` label is applied to all resources related to a
+  particular bot
+
+Note: end users should never need to manually create, modify, or delete these
+objects. All operations on bots and their related objects is performed through
+the `tctl bots` subcommand.
+
+#### `tbot` configuration
+
+The goal is to keep tbot configuration to a minimum to make for a quick getting
+started experience, however we recognize many use cases may quickly become
+difficult to encode on the command line.
+
+To that end, we'd like to follow the philosophy of "keep simple things simple
+but make complex things possible": the simplest / most popular use cases should
+ideally require minimal (zero?) configuration beyond a concise CLI, but more
+complicated use cases can make use of a `tbot.yaml` configuration file.
+
+An example `tbot.yaml` with all default / advanced options specified:
+
+```yaml
+# Storage for the bot's private data
+storage:
+  # can also support any backend supporting both store + fetch
+  directory: /var/lib/teleport/bot
+
+  # alternatively, we can store these certificates only in memory:
+  # memory: {}
+
+# Destinations specify one or more impersonated certificates sets to generate.
+destinations:
+  # `directory` is the only destination supported in the initial iteration;
+  # the tbot process must be given permissions to write to this directory.
+  - directory: /home/alice/tbot
+
+    # One of more roles to grant to the bot. It must have been granted (at
+    # least) these roles with `tctl bots add --roles=...`
+    # By default, all possible roles are included.
+    roles: [a, b, c]
+
+    # Which types of certificates to generate. `ssh` is the default.
+    kinds: [ssh]
+
+    # A list of configuration templates to generate and write to the
+    # destination. Defaults are decided based on the certificate kinds:
+    # `ssh` implies `ssh-client` unless manually (not) configured.
+    configs:
+      # ssh-client generates known_hosts and an ssh_config that can be
+      # included. We can ensure the correct certificate kinds are generated
+      # while generating the config templates.
+      - ssh-client
+
+  - directory: /home/bob/tbot
+    roles: [b, c]
+
+    # TLS certificates can be generated for app / db access.
+    kinds: [ssh, tls]
+
+    configs:
+      # Some future configuration templates may require parameters.
+      # (This is an example; ssh-server may not need this or any parameters)
+      - ssh-server:
+          hostname: example.com
+  
+  # Some backends may optionally accept advanced parameters via an object:
+  - directory:
+      path: /foo/bar
+      symlinks: insecure
+```
+
+In future iterations, additional output destination backends may be considered.
+A few ideas include:
```

Comment:
Would we able to use MachineID to obtain AWS creds via Teleport Application Access?

Reply:
I don't know enough about our AWS app access story today to say for certain, but if AWS can be configured to the trust certs issued by Teleport it should be possible.

This initial iteration of Machine ID only issues barebones certs so at the moment app / db / kubernetes access aren't expected to work. We'll be looking into supporting these as soon as possible, though.